### PR TITLE
[ATen] Set and get default dtype

### PR DESCRIPTION
**Pull request by goldsborough**
Replaces the `DefaultTensorOptions` with just a global default dtype that you can set and get like in Python.

Also, calls `set_default_dtype` in the implementation of `torch.set_default_dtype`. Right now these two default values are separate but will always be the same. Should we just bind `set_default_dtype`  into Python? I think that might be good to do in a separate PR though.

@ezyang @gchanan 

Also CC @colesbury who wanted to do this for ATen for a while? What do you think about it?

### CODE REVIEW

File: aten/src/ATen/core/DefaultDtype.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+
+namespace caffe2 {
+class TypeMeta;
+} // namespace caffe2
+
+namespace at {
+CAFFE2_API void set_default_dtype(caffe2::TypeMeta dtype);
```
Comment (ezyang) [inline aten/src/ATen/core/DefaultDtype.h]: Docs! Docs!

**Comment by ezyang** [comment]
Thanks, this will be easier to reason about.

It seems pretty important to link the Python and C++ sooner rather than later, because I'm still not convinced that all of our internal kernels are necessarily written defensively enough to handle the default dtype having been changed under them. There shouldn't be many, but I bet there are a couple that are just wrong.

**Review by ezyang — APPROVED** [review]
However you feel about the ability to set the default dtype, this is definitely better than what we had before.

**Comment by goldsborough** [comment]
I'm going to keep this PR small and figure out the python convergence in a follow up right after

**Pull request closed without merging**